### ISSUE TITLE
Update function problem on associations has_many mssql / postgresql

### ISSUE THREAD

**Issue by FabriceBazzaro** [sequelize/sequelize#7589]
To begin, thank you for your job, this ORM is really a good one and after using it since months now, it's my first really hard problem. 

PS: Sorry for my english...

## What you are doing?
I update associated objects of an instance I've found by a call of findById:
MyInstance.setAssos([Obj1, Obj2, Obj3]);

## What do you expect to happen?
I want to obtain the three objects as associated objects of my instance

## What is actually happening?
The output is not always the same, sometimes, the association is done correctly and sometimes my instance is without any associated objects

## What I have identified?

What I've found to correct my problem in sequelize : in the association/has_many.js module there is the following problem in the set function : 

At the end of the function:
 - obsoleteAssociations contain all the old Ids
 - unassociatedObjects contain all the new ones

So I have two updates: one delete all the associations and the other create the new associations and with the Promise.all there is a race condition between the two updates and sometimes all is good and sometimes no...

I modified the code to obtain some trace of the execution of the comparison before the generation of updates : 

      obsoleteAssociations = oldAssociations.filter(function(old) {
          return !_.find(targetInstances, function(obj) {
              console.log("--------------------------------------------------------------------------------------------------- < OBJ");
              console.log("TARGET ------->", association.target.primaryKeyAttribute);
              console.log("--------------------------------------------------------------------------------------------------- < OBJ");
              console.log("OBJ ID:", obj[association.target.primaryKeyAttribute], " - TYPE", typeof(obj[association.target.primaryKeyAttribute]));
              console.log("OLD ID:", old[association.target.primaryKeyAttribute], " - TYPE", typeof(old[association.target.primaryKeyAttribute]));
              console.log("=== ID:", obj[association.target.primaryKeyAttribute] === old[association.target.primaryKeyAttribute]);
              console.log("--------------------------------------------------------------------------------------------------- < OBJ");
            return obj[association.target.primaryKeyAttribute] === old[association.target.primaryKeyAttribute];
          });
        })
      , unassociatedObjects = targetInstances.filter(function(obj) {
          return !_.find(oldAssociations, function(old) {
            return obj[association.target.primaryKeyAttribute] === old[association.target.primaryKeyAttribute];
          });
        })

And I obtained the following

--------------------------------------------------------------------------------------------------- < OBJ
TARGET -------> id
--------------------------------------------------------------------------------------------------- < OBJ
OBJ ID: 3211 TYPE string
OLD ID: 3211 TYPE number
=== ID: false
--------------------------------------------------------------------------------------------------- < OBJ

So I've relaxed the equality test with "==" in the two find functions and all is OK for my cases.

__Dialect__: mssql
__Database version:__ SQL Server 2012 and Postgresql
__Sequelize version:__ 3.20.0

**Comment by sushantdhiman** [comment]
> So I have two updates: one delete all the associations and the other create the new associations and with the Promise.all there is a race condition between the two updates and sometimes all is good and sometimes no

I think thats the problem, if you have both associate and unassociate queries in Promise.all they will definitely conflict based on your DBMS response lag.

You should use some serial promise approach like Promise.each / Promise.map etc or even order associate and unassociate in some definite way so your code dont have random effects.

**Comment by FabriceBazzaro** [comment]
I'm agree with you but the this promise is in the Sequelize code, more precisely in the file  "association/has_many.js" of Sequelize. 

The part dedicated of "what I have identified" is in the Sequelize code. Sorry if it wasn't clear, originally, I think the problem is in my code with different types for the foreignkey and for the id field in my two objects (one is coming form a json and the other from a findById) but I don't know if it is important for the project to relax the equality test in this module.

**Comment by sushantdhiman** [comment]
Can you show me the code you use for creating association where you noticed this bug ?

**Comment by FabriceBazzaro** [comment]
I'm confused but the code I use is an implementation of the jsonapi and I have implemented a generator using an abstraction of sequelize. I have a lot of modules linked to understand from where the problem is coming from. I will made further investigations and I'll try to give you some elements asap. 

I have an operation which make some inserts, I use the results of this inserts in an update (on a sequelize instance obtained by a findById) using the set function for the relation (using in parameter the array of inserted objects). 

I've made some verifications on my models, it's standard in its definition:

var description = {
    name: 'Form',
    attributes: {
    },

    associations: {
        hasMany: {
            FormField: {
                as: 'fields',
                foreignKey: 'formId'
            }
...
var description = {
    name: 'FormField',
    attributes: {
    },

    associations: {
        belongsTo: {
            Form: {
                as: 'form',
                foreignKey: 'formId'
            }
        }
    }
...

I've reduced the code, and the object is transmitted to a generator which call the creation of Sequelize models using my definition which have light differences whith the Sequelize description because I use some extra fields.

I can send you my code in mp but It's a lot of code and I don't wan't you to waste your time.
